Suppress key listener when on clickable target

barklund opened the pull request:
## Summary

This fixes the global <kbd>enter</kbd> listener on the canvas to ignore instances where the focus is on a clickable target - in such a case we need browser default behaviour to kick in instead.

Also, this behaviour was manually re-implemented in a number of components using an explicit listener - this has now been removed (leading to a net reduction of code in this PR - yay 🎉 ).

This makes sure that all clickable targets can be correctly clicked using the keyboard - and, in particular, ensures that all WP links (in the menu and topbar) can be correctly navigated.

**NB:** @miina did [suggest this](https://github.com/google/web-stories-wp/pull/2879#discussion_r450094108) when I implemented the feature in the first place, but I somehow assumed it to be harder to do than it was. 🙇  

## Relevant Technical Choices

* Add new `clickable` flag to event listener logic, defaulting to `true`
* If false, ignore clickable targets (buttons, links, some inputs) when assigning the listener

## To-do

* We could potentially add checks for clickable roles (`button`, `option`, `menuitem`), but I'm not sure it's necessary.

## User-facing changes

* WP menu options in can now be navigated to using keyboard

## Testing Instructions

1. Tab to a menu item in the WP left-hand menu.
1. Press enter to navigate to the link target and observe that something actually happens (whereas before, nothing happened)

Fixes #4239

Commit:
Added new clickable flag (default true) for key listeners

Commit:
Removed explicit <enter> handlers across various components

miina commented on assets/src/edit-story/components/keyboard/index.js:
```diff
@@ -39,6 +39,15 @@ const NON_EDITABLE_INPUT_TYPES = [
   'reset',
   'hidden',
 ];
+const CLICKABLE_INPUT_TYPES = [
```
Should the `NON_EDITABLE_INPUT_TYPES` perhaps just be `[ ...CLICKABLE_INPUT_TYPES, 'hidden']`?

barklund replied:
Hm, not sure. Is it an either/or - either an input field is clickable or it's editable (or none of them, `hidden`)? Can that be considered a _universal truth_ about inputs?

barklund replied:
Looking further into this, a counter-example would be `<input type="date" />` - it seems to me to be both _clickable_ and _editable_.

So I'll leave this as-is for now.

miina commented on assets/src/edit-story/components/form/switch.js:
```diff
@@ -124,12 +124,10 @@ function Switch({ value, disabled, onChange, onLabel, offLabel }) {
-    () => handleChange(!value),
-    [handleChange, value]
-  );
+  useKeyDownEffect(ref, ['left', 'right'], () => handleChange(!value), [
```
This now wouldn't change the value on `Enter`/`Space`, is that intended?

barklund replied:
<kbd>space</kbd> already worked even without this handler as it was default behavior. After this fix, <kbd>enter</kbd> also works by browser default behavior. This is according to my tests at least, are you experiencing otherwise?

miina replied:
Correct, neither of these seem to work when testing in Chrome.

barklund replied:
You're right, reverted in 78a23aec4.

miina reviewed (commented):
One thing I noticed when testing:
The bold-italic-underline do not work with `Space` nor `Enter`, also seeing this in console:
<img width="551" alt="Screenshot 2020-08-27 at 14 01 22" src="https://user-images.githubusercontent.com/3294597/91434747-8fcae080-e865-11ea-8acc-f424b0bae0be.png">

Perhaps not related to this PR though.

Commit:
Reverted changes to switch component

Commit:
Restored <enter> listener for checkboxes

Apparently checkboxes are not normally toggled by pressing <enter>, only <space>. <enter> would normally submit the form, like in a text input.

barklund commented on assets/src/edit-story/components/form/toggleButton.js:
```diff
@@ -110,12 +110,8 @@ function ToggleButton({
-  // for text elements. For non-text element selection, this does nothing, that default beviour
-  // wouldn't do.
   const inputRef = useRef();
+  // <enter> doesn't normally toggle checkboxes, but we'd like it to
```
@miina, TIL that checkboxes aren't normally toggled by pressing <kbd>enter</kbd>. But we'd like it to with our buttons (as they're more like _buttons_ than _inputs_), so re-adding the handler here.
